packages/next/server/base-http.ts

```diff
@@ -0,0 +1,287 @@
+import type { ServerResponse, IncomingMessage, IncomingHttpHeaders } from 'http'
+import type { Writable, Readable } from 'stream'
+import { PERMANENT_REDIRECT_STATUS } from '../shared/lib/constants'
+import {
+  getCookieParser,
+  NextApiRequestCookies,
+  parseBody,
+  SYMBOL_CLEARED_COOKIES,
+} from './api-utils'
+import { I18NConfig } from './config-shared'
+import { NEXT_REQUEST_META, RequestMeta } from './request-meta'
+
+export interface BaseNextRequestConfig {
+  basePath: string | undefined
+  i18n?: I18NConfig
+  trailingSlash?: boolean | undefined
+}
+
+export abstract class BaseNextRequest<Body = any> {
+  protected _cookies: NextApiRequestCookies | undefined
+  public abstract headers: IncomingHttpHeaders
+
+  constructor(public method: string, public url: string, public body: Body) {}
+
+  abstract parseBody(limit: string | number): Promise<any>
+
+  // Utils implemented using the abstract methods above
+
+  public get cookies() {
+    if (this._cookies) return this._cookies
+    return (this._cookies = getCookieParser(this.headers)())
+  }
+}
+
+export class NodeNextRequest extends BaseNextRequest<Readable> {
+  public headers = this._req.headers;
+
+  [NEXT_REQUEST_META]: RequestMeta
+
+  get req() {
+    // Need to mimic these changes to the original req object for places where we use it:
+    // render.tsx, api/ssg requests
+    this._req[NEXT_REQUEST_META] = this[NEXT_REQUEST_META]
+    this._req.url = this.url
+    this._req.cookies = this.cookies
+    return this._req
+  }
+
+  constructor(
+    private _req: IncomingMessage & {
+      [NEXT_REQUEST_META]?: RequestMeta
+      cookies?: NextApiRequestCookies
+    }
+  ) {
+    super(_req.method!.toUpperCase(), _req.url!, _req)
+  }
+
+  async parseBody(limit: string | number): Promise<any> {
+    return parseBody(this._req, limit)
+  }
+}
+
+export class WebNextRequest extends BaseNextRequest<ReadableStream | null> {
```

Comment:
(Not blocking) maybe we can separate Web and Node request/response in 2 files just like we did for base-server and web-server.